### PR TITLE
Introduce parallelism in Iceberg `migrate` procedure

### PR DESCRIPTION
**Pull request by krvikash**
## Description
Introduce parallelism in Iceberg `migrate` procedure

## Release notes

(X) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Improve performance when executing `migrate` procedure. ({issue}`16786`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java [hunk 1]
```diff
@@ -309,12 +315,14 @@ public Map<String, Optional<Partition>> listAllPartitions(HiveMetastore metastor
...
     private List<DataFile> buildDataFiles(ConnectorSession session, RecursiveDirectory recursive, HiveStorageFormat format, String location, PartitionSpec partitionSpec, StructLike partition, NameMapping nameMapping)
-            throws IOException
+            throws IOException, ExecutionException, InterruptedException
     {
-        // TODO: Introduce parallelism
         TrinoFileSystem fileSystem = fileSystemFactory.create(session);
         FileIterator files = fileSystem.listFiles(location);
         ImmutableList.Builder<DataFile> dataFilesBuilder = ImmutableList.builder();
+        int numThreads = Runtime.getRuntime().availableProcessors();
+        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java:324]: The thread pool should be Guice-managed.
See example in Hive
https://github.com/trinodb/trino/blob/b6640e96aa4e61f5eb06e9660ace5893aade3b95/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveModule.java#L167-L172
https://github.com/trinodb/trino/blob/2a88bd74e37579484d259f37389c4cb1da5484c8/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java#L139

**Review by findinpath — APPROVED** [review]
@ebyhr  CPTAL?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java [hunk 2]
```diff
@@ -146,12 +152,15 @@ public MigrateProcedure(
             TrinoFileSystemFactory fileSystemFactory,
             TypeManager typeManager,
             IcebergConfig icebergConfig,
-            IcebergSecurityConfig securityConfig)
+            IcebergSecurityConfig securityConfig,
+            ExecutorService coreExecutor)
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java:156]: nit: Move to after `TypeManager typeManager` to use the same order in the fields and reduce the diff.

**Review by ebyhr — APPROVED** [review]

**Comment by krvikash** [comment]
Thanks @findinpath | @ebyhr for the review. I have addressed the comments.

**Review by findinpath — APPROVED** [review]

**Review by alexjo2144 — COMMENTED** [review]
Taking another look at this, I think we should move the parallelism up a bit to be one thread per partition. That'd mean submitting the `buildDataFiles` method to the executor rather than just using it to loadMetrics.

**Comment by mosabua** [comment]
:wave: @krvikash @alexjo2144 @ebyhr - this PR has become inactive. We hope you are still interested in working on it. Please discuss next steps such as rebase and further changes.


We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Pull request closed without merging**